Tasks/Gradle/sonarqube-common/sonarqube-server.ts

```diff
@@ -0,0 +1,112 @@
+import Q = require('q');
+import url = require('url');
+import https = require('https');
+import http = require('http');
+import {IncomingMessage} from 'http';
+
+import {SonarQubeEndpoint} from './sonarqube-common';
+import sqCommon = require('./sonarqube-common');
+
+import tl = require('vsts-task-lib/task');
+
+export interface ISonarQubeServer {
+    callSonarQubeRestEndpoint(path:string):Q.Promise<Object>;
+
+    getEndpoint():SonarQubeEndpoint;
+}
+
+export class SonarQubeServer implements ISonarQubeServer {
+
+    private endpoint:SonarQubeEndpoint;
+
+    constructor() {
+        this.endpoint = sqCommon.getSonarQubeEndpoint();
+    }
+
+    public getEndpoint():SonarQubeEndpoint {
+        return this.endpoint;
+    }
+
+    /**
+     * Invokes a REST endpoint on the SonarQube server.
+     * @param path The host-relative path to the REST endpoint (e.g. /api/ce/task)
+     * @returns A promise, resolving with a JSON object representation of the response. Rejects on error or non-200 header.
+     */
+    public callSonarQubeRestEndpoint(path:string):Q.Promise<Object> {
+        var defer = Q.defer<Object>();
```

Comment:
We have way too much logic in that soanrqube-common.ts file. To break it up, create a separate file for the endpoint and move all the logic there. It's not related to your work, but http://programmer.97things.oreilly.com/wiki/index.php/The_Boy_Scout_Rule